stl/src/xonce2.cpp

```diff
@@ -0,0 +1,30 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// __std_execute_once_begin & __std_execute_once_end functions
+
+#include "xcall_once.h"
+#include <synchapi.h>
+
+// This must be as small as possible, because its contents are
+// injected into the msvcprt.lib and msvcprtd.lib import libraries.
+// Do not include or define anything else here.
+// In particular, basic_string must not be included here.
+
+// these declarations must be in sync with those in xcall_once.h
+
+_STD_BEGIN
+int __CLRCALL_PURE_OR_CDECL __std_execute_once_begin(
+    once_flag& _Once_flag, int& _Pending) noexcept { // wrap Win32 InitOnceBeginInitialize()
+    static_assert(sizeof(_Once_flag._Opaque) == sizeof(INIT_ONCE), "invalid size");
+
+    return ::InitOnceBeginInitialize(reinterpret_cast<PINIT_ONCE>(&_Once_flag._Opaque), 0, &_Pending, nullptr);

```

Comment:
This needs to implement the XP fallback spinlock too

Reply:
Why? I though only `.cpp` files were to maintain compatibility.

Reply:
For new features we can get away without XP support. But this is an existing feature.

Reply:
I mean, this is import library file, like `sharedmutex.cpp`, used only by newly compiled headers

Reply:
The only use of `call_once` by STL itself is here:
https://github.com/microsoft/STL/blob/22391019cba3863a978ea2d7d3f5ca240f38c576/stl/src/ppltasks.cpp#L117

Do you mean that existing STL headers should support XP for inclusion elsewhere?

Reply:
I'm not talking about by the STL itself, I'm talking about by customers. We shouldn't gratuitously break an existing customer program that works today.

> like sharedmutex.cpp

`<shared_mutex>` the whole feature, never supported XP. Today `call_once` does.

Reply:
Ok. But then import library code would be a bit larger for XP import libary build, as it would contain `GetProcAddress`. Is it still acceptable to have it in import library then?

Reply:
Yes, import libraries already GetProcAddress, see `parallel_algorthms.cpp` for example. We should guard such checks for `_STL_WIN32_WINNT < _WIN32_WINNT_VISTA` though so that onecore/xbox/app don't need the GetProcAddress.

Reply:
(Also note that while XP needs to work, the fallback doesn't have to be "good" there, a plain spinlock is fine considering the number of many-core XP machines is basically zero. https://github.com/microsoft/STL/blob/92508bed6387cbdae433fc86279bc446af6f1b1a/stl/src/winapisupp.cpp#L221 is the old fallback; whatever is done here needs to be ABI compatible with that)

Reply:
I found that comment - https://github.com/microsoft/STL/issues/606#issuecomment-599739740:

> we can now assume that XP is never targeted by the headers, we just have to continue supporting it in the redist

Reply:
See, it would be very helpful to confirm whether headers should maintain XP compatibility or not.
If not, it would be easy to remove exception pointer forwarding from `<mutex>` and improve thorughput.

I see I can implement substitute `InitOnceBeginInitialize` /  `InitOnceComplete` and never have old flow, even in XP, but I'm afraid this path will not be well tested.

Reply:
I just asked around about this for confirmation from the business side. If business side is OK with dropping XP then we can use symbol aliases and avoid a whole extra API hop here. Will keep you posted.

Reply:
If the answer is negative, I'm leaning towards still cleaning header from old flow.
Then for XP I can implement `InitOnceBeginInitialize` and `InitOnceComplete` fallbacks compatible with existing `InitOnceExecuteOnce` fallback (same meaning of `once_flag` values). With #593 arrival, the new XP fallback can reuse timed backoff spinning for XP (which is already there in current #593).

Reply:
Any updates on that? 
For now I've duplicated spin wait from `winapisupp.cpp`, enhancing it with spinning duplication from #593 , as cleaning up `<mutex>` looks too attractive to miss it even if going to support XP.